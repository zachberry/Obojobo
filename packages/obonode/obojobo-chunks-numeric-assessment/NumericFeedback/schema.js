import { Block } from 'slate'

import SchemaViolations from 'obojobo-document-engine/src/scripts/oboeditor/util/schema-violations'
import { NUMERIC_FEEDBACK_NODE } from '../constants'

const { CHILD_TYPE_INVALID, CHILD_MIN_INVALID } = SchemaViolations

const TEXT_NODE = 'ObojoboDraft.Chunks.Text'
const PAGE_NODE = 'ObojoboDraft.Pages.Page'

const schema = {
	blocks: {
		[NUMERIC_FEEDBACK_NODE]: {
			nodes: [{ match: [{ type: 'oboeditor.component' }], min: 1 }],
			normalize: (editor, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_MIN_INVALID: {
						const block = Block.create({
							object: 'block',
							type: 'oboeditor.component',
							nodes: [
								{
									object: 'block',
									type: PAGE_NODE,
									nodes: [{ object: 'block', type: TEXT_NODE }]
								}
							]
						})
						return editor.insertNodeByKey(node.key, index, block)
					}
					case CHILD_TYPE_INVALID: {
						const block = Block.fromJSON({
							object: 'block',
							type: 'oboeditor.component',
							nodes: [
								{
									object: 'block',
									type: PAGE_NODE,
									nodes: [{ object: 'block', type: TEXT_NODE }]
								}
							]
						})
						return editor.withoutNormalizing(c => {
							c.removeNodeByKey(child.key)
							return c.insertNodeByKey(node.key, index, block)
						})
					}
				}
			}
		}
	}
}

export default schema
