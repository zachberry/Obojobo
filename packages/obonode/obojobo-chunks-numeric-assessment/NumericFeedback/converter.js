import Component from 'obojobo-document-engine/src/scripts/oboeditor/components/node/editor'

const slateToObo = node => ({
	id: node.key,
	type: node.type,
	children: node.nodes.map(child => Component.helpers.slateToObo(child)),
	content: node.data.get('content') || {}
})

const oboToSlate = node => ({
	object: 'block',
	key: node.id,
	type: node.type,
	nodes: node.children.map(child => Component.helpers.oboToSlate(child)),
	data: {
		content: node.content
	}
})

export default { slateToObo, oboToSlate }
