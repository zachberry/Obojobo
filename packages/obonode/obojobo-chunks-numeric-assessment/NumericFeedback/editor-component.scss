@import '~styles/includes';

.numeric-feedback-editor {
	$color-border: #dfdfdf;

	position: relative;
	margin-top: 1rem;
	padding: 1.1rem 0;

	border: 1px solid $color-border;
	background: $color-bg;

	&:before {
		color: $color-text-minor;
		content: 'Feedback';
		display: block;
		font-family: $font-default;
		position: absolute;
		left: 3.65rem;
		top: 0.5rem;
		transform: scale(0.7);
		transform-origin: top left;
	}

	.feedback-delete-button {
		background-color: $color-bg;
		border: $color-bg;
		right: 0.1rem;
		top: 0.2rem;
		position: absolute;

		button {
			pointer-events: none;
		}
	}
}
