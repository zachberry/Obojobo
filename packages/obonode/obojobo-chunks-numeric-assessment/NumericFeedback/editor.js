import React from 'react'
import Common from 'obojobo-document-engine/src/scripts/common'

import Node from './editor-component'
import Schema from './schema'
import Converter from './converter'

import { NUMERIC_FEEDBACK_NODE } from '../constants'

const plugins = {
	renderNode(props, editor, next) {
		switch (props.node.type) {
			case NUMERIC_FEEDBACK_NODE:
				return <Node {...props} {...props.attributes} />
			default:
				return next()
		}
	},
	schema: Schema
}

Common.Registry.registerModel(NUMERIC_FEEDBACK_NODE, {
	name: 'Numeric Choice Feedback',
	isInsertable: false,
	slateToObo: Converter.slateToObo,
	oboToSlate: Converter.oboToSlate,
	supportsChildren: true,
	plugins
})

const NumericFeedback = {
	components: {
		Node
	},
	helpers: {
		slateToObo: Converter.slateToObo,
		oboToSlate: Converter.oboToSlate
	},
	plugins
}

export default NumericFeedback
