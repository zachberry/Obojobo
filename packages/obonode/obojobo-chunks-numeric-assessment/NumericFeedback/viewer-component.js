import './viewer-component.scss'

import React from 'react'
import Viewer from 'obojobo-document-engine/src/scripts/viewer'

const { OboComponent } = Viewer.components

const MCFeedback = props => (
	<OboComponent
		model={props.model}
		moduleData={props.moduleData}
		className={`obojobo-draft--chunks--numeric-assessment--numeric-feedback${
			3
			// props.model.parent.modelState.score === 100
			// 	? ' is-correct-feedback'
			// 	: ' is-not-correct-feedback'
		}`}
	>
		<span className="for-screen-reader-only">. Feedback for your answer:</span>
		{props.model.children.models.map(child => {
			const Component = child.getComponentClass()
			return <Component key={child.get('id')} model={child} moduleData={props.moduleData} />
		})}
	</OboComponent>
)

export default MCFeedback
