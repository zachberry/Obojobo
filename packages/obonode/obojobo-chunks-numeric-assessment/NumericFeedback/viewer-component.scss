@import '~styles/includes';

.obojobo-draft--chunks--numeric-assessment--numeric-feedback {
	font-size: 0.8em;
	padding-left: $dimension-padding * 0.2333333333;
	padding-right: $dimension-padding * 0.2333333333;

	&.editor-feedback {
		position: relative;
		margin-top: 2em;
		padding-top: 2em;
		padding-bottom: 2em;
		pointer-events: all;
		background: $color-bg2;
	}

	.label {
		position: absolute;
		top: -0.1em;
		left: 0.6em;
		color: $color-text-minor;
		transform: scale(0.7);
		transform-origin: top left;
		font-family: $font-default;
	}
}
