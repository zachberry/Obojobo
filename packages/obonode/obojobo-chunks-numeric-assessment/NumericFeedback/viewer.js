import Common from 'obojobo-document-engine/src/scripts/common'
import ViewerComponent from './viewer-component'
import { NUMERIC_FEEDBACK_NODE } from '../constants'

Common.Registry.registerModel(NUMERIC_FEEDBACK_NODE, {
	adapter: null,
	componentClass: ViewerComponent,
	type: 'chunk'
})
