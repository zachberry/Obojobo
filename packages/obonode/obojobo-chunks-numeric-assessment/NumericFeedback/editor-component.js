import './viewer-component.scss'
import './editor-component.scss'

import React from 'react'

const NumericFeedback = props => {
	const onDelete = () => props.editor.removeNodeByKey(props.node.key)

	return (
		<div className="component numeric-feedback-editor" contentEditable={true}>
			<button className="feedback-delete-button" onClick={onDelete} contentEditable={false}>
				Ã—
			</button>
			{props.children}
		</div>
	)
}

export default NumericFeedback
