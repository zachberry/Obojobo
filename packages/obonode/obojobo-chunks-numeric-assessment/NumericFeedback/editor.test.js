import { CHILD_TYPE_INVALID } from 'slate-schema-violations'

import NumericFeedback from './editor'
import { NUMERIC_FEEDBACK_NODE } from '../constants'

describe('NumericFeedback editor', () => {
	test('plugins.renderNode renders a node', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: NUMERIC_FEEDBACK_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(NumericFeedback.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode calls next', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: 'mockNode',
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		const next = jest.fn()

		expect(NumericFeedback.plugins.renderNode(props, null, next)).toMatchSnapshot()
		expect(next).toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes invalid children', () => {
		const editor = {
			removeNodeByKey: jest.fn(),
			insertNodeByKey: jest.fn()
		}

		editor.withoutNormalizing = jest.fn().mockImplementationOnce(funct => {
			funct(editor)
		})

		NumericFeedback.plugins.schema.blocks[NUMERIC_FEEDBACK_NODE].normalize(editor, {
			code: CHILD_TYPE_INVALID,
			node: {},
			child: { key: 'mockKey', object: 'text' },
			index: null
		})

		expect(editor.insertNodeByKey).toHaveBeenCalled()
	})

	test('plugins.schema.normalize adds missing children', () => {
		const editor = {
			insertNodeByKey: jest.fn()
		}

		NumericFeedback.plugins.schema.blocks[NUMERIC_FEEDBACK_NODE].normalize(editor, {
			code: 'child_min_invalid',
			node: {},
			child: null,
			index: 0
		})

		expect(editor.insertNodeByKey).toHaveBeenCalled()
	})
})
