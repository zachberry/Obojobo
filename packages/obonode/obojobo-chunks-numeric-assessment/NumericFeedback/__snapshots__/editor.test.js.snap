// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericFeedback editor plugins.renderNode calls next 1`] = `undefined`;

exports[`NumericFeedback editor plugins.renderNode renders a node 1`] = `
<NumericFeedback
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.NumericAssessment.NumericFeedback",
    }
  }
/>
`;
